core/src/server/rpc_module.rs

```diff
@@ -806,27 +896,25 @@ impl SubscriptionSink {
 					Either::Left((Ok(Some(result)), next_closed_fut)) => {
 						match self.send(&result) {
 							Ok(_) => (),
-							Err(Error::SubscriptionClosed(close_reason)) => {
-								self.close(&close_reason);
-								break Ok(());
-							}
 							Err(err) => {
 								break Err(err);
 							}
 						};
 						stream_item = stream.try_next();
 						closed_fut = next_closed_fut;
 					}
+					// Stream canceled because of error.
 					Either::Left((Err(e), _)) => {
-						let close_reason = SubscriptionClosedReason::Server(e.to_string()).into();
-						self.close(&close_reason);
+						let close_reason: SubscriptionClosed = CloseReason::Failed(e.to_string()).into();
+						break Err(Error::SubscriptionClosed(close_reason));
+					}
+					// Stream completed (this is not really an error)
+					Either::Left((Ok(None), _)) => {
+						let close_reason: SubscriptionClosed = CloseReason::Success.into();
 						break Err(Error::SubscriptionClosed(close_reason));
```

Comment:
If my understanding in the above comment is correct though, should we just return `Ok(())` here, since we've made no assumptions about how the subscription communicates that it's been closed (ie it may have already sent something via the "ok" channel)?

Reply:
I changed this look at the latest changes, should be according to the PR description now.

Changed from it  to Result<SubscriptionResult, Error>` but we just have have `SubscriptionClosed` there and remove the Result entirely for simplicity.